### PR TITLE
Initial work on the java module.

### PR DESCRIPTION
**Pull request by internettrans**
This PR creates a java omod module that is put into `~/.OpenMRS/modules` and gets Tomcat to always return a single HTML file for all frontend routes, instead of expecting there to be a JSP for each url. See the README.md within this PR for more details.

The vast majority of the files in this PR were created via `openmrs-sdk:create-project`, and as such I don't really understand everything in them or their purpose. See https://wiki.openmrs.org/display/docs/Creating+Modules and https://wiki.openmrs.org/display/docs/OpenMRS+SDK#OpenMRSSDK-Creatingprojects for documentation on it.

I have added comments to all of the files that I think are relevant for a code review.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,29 @@
-# openmrs-frontend-spa
-The frontend SPA for OpenMRS code
+# openmrs-module-spa
```
Comment (internettrans) [inline README.md:1]: This file is worth reading over and reviewing. Especially `4.`

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.InputStream;
+import java.util.Scanner;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class SinglePageApplicationServlet extends HttpServlet {
```
Comment (internettrans) [inline omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:20]: I wrote this entire file from scratch. Worth reviewing. The purpose of this class is to always return the index.htm file with an HTTP status of 200, since the frontend is going to handle routing. I don't know OpenMRS best practices for Java code, and it's been a few years since I've written Java code so please give suggestions if you find problems with it.

File: omod/src/main/resources/META-INF/web-fragment.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (internettrans) [inline omod/src/main/resources/META-INF/web-fragment.xml]: I wrote this file and am hoping to find a way to get this file to actually be integrated into OpenMRS' web.xml file without modifying openmrs-core. See related https://www.ibm.com/support/knowledgecenter/SS7K4U_9.0.0/com.ibm.websphere.zseries.doc/ae/cweb_webfragments.html and https://www.oracle.com/technetwork/articles/javaee/javaee6overview-part2-136353.html.

This file is related to the issue described in `4.` in the Readme.

Unfortunately, I have been unable to get Tomcat/Jetty to respect this web fragment config.

File: omod/src/main/webapp/index.htm [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
```
Comment (internettrans) [inline omod/src/main/webapp/index.htm:1]: This file is worth reviewing. It is the index.html that will be the SPA for all open mrs code.

It's very barebones and has nothing in it yet, but subsequent PRs will add to it.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,29 @@
...
+
+This project is an OpenMRS java module that makes it possible for Tomcat to serve an SPA instead of expecting all routes
+to be handled by server side JSPs/GSPs.
+
+It is part of the microfrotends architecture that is explained at https://github.com/openmrs/openmrs-rfc-frontend.
+
+## Development instructions
```
Comment (internettrans) [inline README.md:8]: Don't worry -- in the future, we won't have to follow all of the steps in here every time we want to write frontend code.

Once we get a couple key things in place, we'll be able to boot up a webpack-dev-server, create-react-app, `ng serve`, etc and not even have to bother with booting up a backend server because we'll override a deployed environment to point to our localhost for the js assets we're working on. For more info on how that will work, check out https://github.com/joeldenning/import-map-overrides/ and http://single-spa-playground.org/playground/html-file

**Comment by internettrans** [comment]
@brandones @fatmali @angshu I would appreciate your review on this!

File: .project [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>spa</name>
```
Comment (dkayiwa) [inline .project]: You can add these eclipse specific files to the list that is to be ignored. That way, we do not commit them.

File: .settings/org.eclipse.core.resources.prefs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline .settings/org.eclipse.core.resources.prefs]: Same as above

File: .settings/org.eclipse.m2e.core.prefs [hunk 8]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline .settings/org.eclipse.m2e.core.prefs]: Same as above

File: api/.classpath [hunk 9]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" output="target/classes" path="src/main/java">
+		<attributes>
+			<attribute name="optional" value="true"/>
+			<attribute name="maven.pomderived" value="true"/>
```
Comment (dkayiwa) [inline api/.classpath]: Same as above

File: api/.project [hunk 10]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>spa-api</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
```
Comment (dkayiwa) [inline api/.project]: Same as above

File: api/.settings/org.eclipse.core.resources.prefs [hunk 11]
```diff
@@ -0,0 +1,6 @@
+eclipse.preferences.version=1
+encoding//src/main/java=UTF-8
+encoding//src/main/resources=UTF-8
```
Comment (dkayiwa) [inline api/.settings/org.eclipse.core.resources.prefs]: Same as above

File: api/.settings/org.eclipse.jdt.apt.core.prefs [hunk 12]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline api/.settings/org.eclipse.jdt.apt.core.prefs]: Same as above

File: api/.settings/org.eclipse.jdt.core.prefs [hunk 13]
```diff
@@ -0,0 +1,9 @@
+eclipse.preferences.version=1
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.6
+org.eclipse.jdt.core.compiler.compliance=1.6
+org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
```
Comment (dkayiwa) [inline api/.settings/org.eclipse.jdt.core.prefs]: Same as above

File: api/.settings/org.eclipse.m2e.core.prefs [hunk 14]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline api/.settings/org.eclipse.m2e.core.prefs]: Same as above

File: omod/.classpath [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+<classpath>
+	<classpathentry kind="src" output="target/classes" path="src/main/java">
+		<attributes>
+			<attribute name="optional" value="true"/>
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry excluding="**" kind="src" output="target/classes" path="src/main/resources">
+		<attributes>
+			<attribute name="maven.pomderived" value="true"/>
```
Comment (dkayiwa) [inline omod/.classpath]: Same as above

File: omod/.project [hunk 16]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>spa-omod</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
+		<buildCommand>
```
Comment (dkayiwa) [inline omod/.project]: Same as above

File: omod/.settings/org.eclipse.jdt.apt.core.prefs [hunk 17]
```diff
@@ -0,0 +1,2 @@
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline omod/.settings/org.eclipse.jdt.apt.core.prefs]: Same as above

File: omod/.settings/org.eclipse.jdt.core.prefs [hunk 18]
```diff
@@ -0,0 +1,9 @@
+eclipse.preferences.version=1
+org.eclipse.jdt.core.compiler.codegen.targetPlatform=1.6
+org.eclipse.jdt.core.compiler.compliance=1.6
+org.eclipse.jdt.core.compiler.problem.enablePreviewFeatures=disabled
```
Comment (dkayiwa) [inline omod/.settings/org.eclipse.jdt.core.prefs]: Same as above

File: omod/.settings/org.eclipse.m2e.core.prefs [hunk 19]
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
+eclipse.preferences.version=1
```
Comment (dkayiwa) [inline omod/.settings/org.eclipse.m2e.core.prefs]: Same as above

File: omod/src/main/java/org/openmrs/module/spa/web/controller/FrontendSPAController.java [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * This class configured as controller using annotation and mapped with the URL of
+ * 'module/${rootArtifactid}/${rootArtifactid}Link.form'.
+ */
+@Controller("${rootrootArtifactid}.FrontendSPAController")
+@RequestMapping(value = "module/${rootArtifactid}/${rootArtifactid}.form")
+public class FrontendSPAController {
+	
+	/** Logger for this class and subclasses */
+	protected final Log log = LogFactory.getLog(getClass());
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/spa/web/controller/FrontendSPAController.java]: This is our convention for logging: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-loggingLogging

File: omod/src/main/resources/META-INF/web-fragment.xml (hunk 3 above)
Comment (dkayiwa) [reply omod/src/main/resources/META-INF/web-fragment.xml]: Is this of help? https://wiki.openmrs.org/display/docs/Module+Servlets

File: README.md [hunk 21]
```diff
@@ -1,2 +1,29 @@
...
+1. Clone this repository.
+2. `mvn clean install`
+3. Copy `omod/target/spa-1.0.0-SNAPSHOT.jar` to the `~/.OpenMRS/modules` directory. On mac/linux, this is `cp omod/target/spa-1.0.0-SNAPSHOT.omod ~/.OpenMRS/modules`
+4. Add the following to [the web.xml file in openmrs-core](https://github.com/openmrs/openmrs-core/blob/master/webapp/src/main/webapp/WEB-INF/web.xml).
```
Comment (brandones) [inline README.md:12]: Let's ask around about this (after merging).

File: .project (hunk 6 above)
Comment (brandones) [reply .project]: Yeah let's maybe add https://github.com/github/gitignore/blob/master/Global/Eclipse.gitignore to the gitignore? Some of these files are portable, some are not.

File: api/src/main/java/org/openmrs/module/spa/Item.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+/**
+ * Please note that a corresponding table schema must be created in liquibase.xml.
+ */
+//Uncomment 2 lines below if you want to make the Item class persistable, see also FrontendSPADaoTest and liquibase.xml
+//@Entity(name = "spa.Item")
+//@Table(name = "spa_item")
+public class Item extends BaseOpenmrsData {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/spa/Item.java]: What's all this for? Maybe I'm lacking some basic OpenMRS knowledge here, but maybe it could be briefly explained in the class documentation?

File: api/src/main/java/org/openmrs/module/spa/FrontendSPAConfig.java [hunk 23]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/spa/FrontendSPAConfig.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/java/org/openmrs/module/spa/Item.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
+/**
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/spa/Item.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/java/org/openmrs/module/spa/api/FrontendSPAService.java [hunk 25]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/spa/api/FrontendSPAService.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/java/org/openmrs/module/spa/api/dao/FrontendSPADao.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
+/**
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/spa/api/dao/FrontendSPADao.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/java/org/openmrs/module/spa/api/impl/FrontendSPAServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,52 @@
+/**
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/spa/api/impl/FrontendSPAServiceImpl.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/java/org/openmrs/module/spa/Item.java (hunk 22 above)
Comment (brandones) [reply api/src/main/java/org/openmrs/module/spa/Item.java]: Oh wait, I imagine this is OpenMRS module boilerplate... don't worry about it.

File: api/src/main/resources/liquibase.xml [hunk 28]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mseaton) [inline api/src/main/resources/liquibase.xml]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/resources/messages.properties [hunk 29]
```diff
@@ -0,0 +1 @@
+spa.title=Frontend SPA
```
Comment (mseaton) [inline api/src/main/resources/messages.properties]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/resources/messages_es.properties [hunk 30]
```diff
@@ -0,0 +1 @@
+spa.title=Frontend SPA
```
Comment (mseaton) [inline api/src/main/resources/messages_es.properties]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/resources/messages_fr.properties [hunk 31]
```diff
@@ -0,0 +1 @@
+spa.title=Frontend SPA
```
Comment (mseaton) [inline api/src/main/resources/messages_fr.properties]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/main/resources/moduleApplicationContext.xml [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+  		    http://www.springframework.org/schema/tx
+  		    http://www.springframework.org/schema/tx/spring-tx-3.0.xsd
+  		    http://www.springframework.org/schema/aop
+  		    http://www.springframework.org/schema/aop/spring-aop-3.0.xsd
+  		    http://www.springframework.org/schema/util
+  		    http://www.springframework.org/schema/util/spring-util-3.0.xsd">
+
+    <context:component-scan base-package="org.openmrs.module.spa" />
+
+    <!-- Wraps FrontendSPAService methods in DB transactions and OpenMRS interceptors,
```
Comment (mseaton) [inline api/src/main/resources/moduleApplicationContext.xml]: All of the below bean code should be removed, since these classes should be removed.

File: api/src/test/java/org/openmrs/module/spa/FrontendSPAServiceTest.java [hunk 33]
```diff
@@ -0,0 +1,63 @@
+/**
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/module/spa/FrontendSPAServiceTest.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: api/src/test/java/org/openmrs/module/spa/dao/FrontendSPADaoTest.java [hunk 34]
```diff
@@ -0,0 +1,58 @@
+/**
```
Comment (mseaton) [inline api/src/test/java/org/openmrs/module/spa/dao/FrontendSPADaoTest.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: omod/src/main/java/org/openmrs/module/spa/extension/html/AdminList.java [hunk 35]
```diff
@@ -0,0 +1,50 @@
+/**
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/spa/extension/html/AdminList.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:21]: Could a Spring Web Controller accomplish the same thing without the need for a custom servlet?

File: omod/src/main/java/org/openmrs/module/spa/web/controller/FrontendSPAController.java [hunk 37]
```diff
@@ -0,0 +1,87 @@
+/**
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/spa/web/controller/FrontendSPAController.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.  Was this class created deliberately?

File: omod/src/main/resources/config.xml [hunk 38]
```diff
@@ -0,0 +1,106 @@
...
+
+	<activator>org.openmrs.module.spa.FrontendSPAActivator</activator>
+	
+	<!-- <updateURL>https://modules.openmrs.org/modules/download/@MODULE_ID@/update.rdf</updateURL> -->
+	<!-- /Base Module Properties -->
+	
+	<require_version>1.11.6</require_version>
+	
+	<!-- Extensions -->
+	<extension>
```
Comment (mseaton) [inline omod/src/main/resources/config.xml]: This extension should be removed (along with the class it refers to).

File: omod/src/main/resources/config.xml [hunk 39]
```diff
@@ -0,0 +1,106 @@
...
+		<point>org.openmrs.admin.list</point>
+		<class>org.openmrs.module.spa.extension.html.AdminList</class>
+	</extension>
+
+	<aware_of_modules>
```
Comment (mseaton) [inline omod/src/main/resources/config.xml]: Does this need to be aware of legacyui?

File: omod/src/main/resources/config.xml [hunk 40]
```diff
@@ -0,0 +1,106 @@
...
+		<aware_of_module>org.openmrs.module.legacyui</aware_of_module>
+	</aware_of_modules>
+	
+
```
Comment (mseaton) [inline omod/src/main/resources/config.xml]: Everything below this comment should be removed from the file.  It is either incorrect or commented out.

File: omod/src/test/java/org/openmrs/module/spa/AdminListExtensionTest.java [hunk 41]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (mseaton) [inline omod/src/test/java/org/openmrs/module/spa/AdminListExtensionTest.java]: Please remove all classes that are not explicitly needed.  The SDK creates a bunch of "helpful" base classes for a common use case, but these should be deleted before committing if they are not needed.

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java [hunk 42]
```diff
@@ -0,0 +1,53 @@
...
+	
+	private Log log = LogFactory.getLog(this.getClass());
+	
+	public SinglePageApplicationServlet() {
+		super();
+		try {
+			InputStream inputStream = SinglePageApplicationServlet.class.getResourceAsStream("/web/module/index.htm");
+			Scanner scanner = new Scanner(inputStream);
+			scanner.useDelimiter("\\Z");
+			this.htmlFile = scanner.next();
```
Comment (brandones) [inline omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:32]: So it seems like if this fails, you're returning an empty document with a 200?

Maybe we should initialize htmlFile to null (just `private String htmlfile;`), do a null check in `doGet`, and if null return a 500?

**Review by mseaton — CHANGES_REQUESTED** [review]
@joeldenning most of my comments are related to the fact that the SDK or some other utility was used to create this module, and it created a bunch of boilerplate classes and files that are intended to get new developers up and running with some core constructs, but which we have no need for in this module.  Let's pare this down to the bare bones essentials in order to keep ourselves focused on what this module actually does, otherwise things get muddy.  Thanks!

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java [hunk 43]
```diff
@@ -0,0 +1,53 @@
...
+	protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+		response.addHeader("content-type", "text/html");
+		response.setStatus(HttpServletResponse.SC_OK);
+		
+		try {
+			response.getWriter().write(this.htmlFile);
+			response.getWriter().flush();
+		}
+		catch (Exception e) {
+			log.error(e);
```
Comment (brandones) [inline omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:50]: Would this yield a 200 with an empty response? Or fail to respond?

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 36 above)
Comment (mseaton) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:21]: If the goal of this class is to just to ensure a particular page is returned at the root, you can accomplish this with a Spring Controller.  See what we do in our pihmalawi module here:
https://github.com/PIH/openmrs-module-pihmalawi/blob/master/omod/src/main/java/org/openmrs/module/pihmalawi/HomepageOverrideController.java
@joeldenning  / @dkayiwa

File: README.md [hunk 44]
```diff
@@ -1,2 +1,29 @@
...
+
+This project is an OpenMRS java module that makes it possible for Tomcat to serve an SPA instead of expecting all routes
+to be handled by server side JSPs/GSPs.
+
+It is part of the microfrotends architecture that is explained at https://github.com/openmrs/openmrs-rfc-frontend.
+
+## Development instructions
+1. Clone this repository.
+2. `mvn clean install`
+3. Copy `omod/target/spa-1.0.0-SNAPSHOT.jar` to the `~/.OpenMRS/modules` directory. On mac/linux, this is `cp omod/target/spa-1.0.0-SNAPSHOT.omod ~/.OpenMRS/modules`
```
Comment (mogoodrich) [inline README.md:11]: If you are using the SDK you should also be able to do a "openmrs-sdk:watch" on this module and then when you run it will compile this module and then copy it into the modules directory for you. 

You can copy it manually like you say here to avoid having to recompile every time, but really you should be copying it into the modules directory from wheever your current OpenMRS instance is running (the '~/.OpenMRS/modules' assumes you are running a single instace via Tomcat)

File: README.md [hunk 45]
```diff
@@ -1,2 +1,29 @@
...
+    <servlet-class>org.openmrs.module.spa.web.SinglePageApplicationServlet</servlet-class>
+</servlet>
+
+<servlet-mapping>
+    <servlet-name>default</servlet-name>
+    <url-pattern>/</url-pattern>
+</servlet-mapping>
+```
+5. Run openmrs-core locally. I have been using `mvn jetty:run` instead of `mdn openmrs-sdk:run` because it uses my locally
+changed files.
```
Comment (mogoodrich) [inline README.md:28]: If you run a "mvn openmrs-sdk:watch" from the directory where you have core checked out, the SDK should build run at runtime and use your locally-built version.

File: README.md [hunk 46]
```diff
@@ -1,2 +1,29 @@
...
+```xml
+<servlet>
+    <servlet-name>default</servlet-name>
+    <servlet-class>org.openmrs.module.spa.web.SinglePageApplicationServlet</servlet-class>
+</servlet>
+
+<servlet-mapping>
+    <servlet-name>default</servlet-name>
+    <url-pattern>/</url-pattern>
+</servlet-mapping>
```
Comment (mogoodrich) [inline README.md:25]: @mseaton  @brandones  and I were talking about this today... in some ways it would be good not to touch core, but, in this case, since this is designed be such a fundamental feature of OpenMRS going forward, I'd be in support of changing Core if need be... from a PIH EMR perspective, we have finally gotten ourselves on a recent version of OpenMRS Core, so upgrading should be much work.  Would be good to check with AMPATH and Bahmni.

Also, if it turns out we are going to need to modify core regardless, it would likely make sense to *not* create a new module but instead just incorporate all the changes in this module into core.

So, in summary, if it's easy enough to set this all up in a module, let's start with that, but if we end up having to go through a lot of hoops to get things working, we should consider just modifying core.

**Review by mogoodrich — COMMENTED** [review]
I've added my feedback... thanks to @mseaton  for flagging all the boilerplate code that can be removed... for reference, SinglePageApplicationServlet is really the only new class we are adding at this point, correct?

File: README.md (hunk 21 above)
Comment (internettrans) [reply README.md:12]: I think the suggestion from @djazayeri will maybe fix this. I'm going to try it out now. https://github.com/openmrs/openmrs-module-spa/pull/1#discussion_r283107536

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 43 above)
Comment (internettrans) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:50]: Fail to respond -- I'm not sure what we should do beyond just logging the error. Any suggestions?

File: README.md (hunk 46 above)
Comment (internettrans) [reply README.md:25]: 👍 yeah I agree. I was thinking that I'd start it out as a separate module because that seemed less intrusive or likely to break existing code, but then as we grow more confident in it we can decide whether to just bring it into core.

File: README.md (hunk 45 above)
Comment (internettrans) [reply README.md:28]: 👍 thanks for the tip

File: README.md (hunk 44 above)
Comment (internettrans) [reply README.md:11]: 👍

File: .project (hunk 6 above)
Comment (internettrans) [reply .project]: 👍 thanks for the link -- added that to the gitignore and removed the files

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 36 above)
Comment (internettrans) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:21]: @mseaton It's been a while since I've done Spring, but it looks like you have to `@RequestMapping` on each file individually, which wouldn't work for an SPA since any arbitrary frontend route needs to return the index.html file, regardless of the URL.

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 42 above)
Comment (internettrans) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:32]: I like it -- will change.

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 43 above)
Comment (brandones) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:50]: Returning 500?

Comment (internettrans) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:50]: 👍 yeah i saw your other comment -- working on changing this now to return a 500 and log as an error

File: omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java (hunk 36 above)
Comment (mseaton) [reply omod/src/main/java/org/openmrs/module/spa/web/SinglePageApplicationServlet.java:21]: @joeldenning yes, sorry I should have looked closer - I realize now that you are trying to handle every url that isn't already handled.

**Comment by internettrans** [comment]
Closing this because after talking with Daniel we realized that this isn't going to be possible without modifying openmrs-core. So we're moving all of the logic into there, instead of having it in openmrs-module-spa.

Not sure if we'll end up needing openmrs-module-spa or not, so I'll leave this repo extant for a bit. It might turn into a repo that builds frontend modules and then gets them into a jar/omod/maven.
